iavl_versioned_tree.go

```diff
@@ -0,0 +1,167 @@
+package iavl
+
+import (
+	"fmt"
+	"github.com/pkg/errors"
+	dbm "github.com/tendermint/tmlibs/db"
+)
+
+var ErrVersionDoesNotExist = fmt.Errorf("version does not exist")
+
+// VersionedTree is a persistent tree which keeps track of versions.
+type VersionedTree struct {
+	*orphaningTree                           // The current, working tree.
+	versions       map[uint64]*orphaningTree // The previous, saved versions of the tree.
+	latestVersion  uint64                    // The latest saved version.
+	ndb            *nodeDB
+}
+
+// NewVersionedTree returns a new tree with the specified cache size and datastore.
+func NewVersionedTree(cacheSize int, db dbm.DB) *VersionedTree {
+	ndb := newNodeDB(cacheSize, db)
+	head := &IAVLTree{ndb: ndb}
+
+	return &VersionedTree{
+		orphaningTree: newOrphaningTree(head),
+		versions:      map[uint64]*orphaningTree{},
+		ndb:           ndb,
+	}
+}
+
+// LatestVersion returns the latest saved version of the tree.
+func (tree *VersionedTree) LatestVersion() uint64 {
+	return tree.latestVersion
+}
+
+// Tree returns the current working tree.
+func (tree *VersionedTree) Tree() *IAVLTree {
+	return tree.orphaningTree.IAVLTree
+}
+
+// String returns a string representation of the tree.
+func (tree *VersionedTree) String() string {
+	return tree.ndb.String()
+}
+
+// Load a versioned tree from disk. All tree versions are loaded automatically.
+func (tree *VersionedTree) Load() error {
+	roots, err := tree.ndb.getRoots()
+	if err != nil {
+		return err
+	}
+
+	// Load all roots from the database.
+	for _, root := range roots {
+		t := newOrphaningTree(&IAVLTree{ndb: tree.ndb})
+		t.Load(root)
+
+		version := t.root.version
+		tree.versions[version] = t
+
+		if version > tree.latestVersion {
+			tree.latestVersion = version
+		}
+	}
+	// Set the working tree to a copy of the latest.
+	tree.orphaningTree = tree.versions[tree.latestVersion].Clone()
+
+	return nil
+}
+
+// GetVersioned gets the value at the specified key and version.
+func (tree *VersionedTree) GetVersioned(key []byte, version uint64) (
```

Comment:
I think we could just call this `Get`. Alternatively, we could use a `tree.Get(version).Get(key)` pattern. Thoughts ?

Reply:
That's one option for sure, then we'd have to rename `Get` to `GetLatest`. The reason I have it this way is that `Get` has the same type for both versioned and non-versioned trees. I'm not a fan of the double get, I'd rather expose only one interface (VersionedTree).